### ISSUE TITLE
Getting Unknown symbol __aeabi_uldivmod on latest

### ISSUE THREAD

**Issue by ewildgoose** [vel21ripn/nDPI#110]
Hi, latest flow-3.2 is giving me errors about "Unknown symbol __aeabi_uldivmod" again. I'm fairly sure this is an attempt to do 64 bit division in kernel space, which isn't allowed on 32bit kernels. My specific board under test is arm, but I suspect I'll get the same on my Alix 32bit x86 board as well?

Any thoughts before I try and debug this further? In the past I think it's needed patching for the flow counter stats. There is a do_div() macro that can be used to kernel space to do the division, we just need to carry the patch forward on version bumps?

It's such a pain finding these. If anyone has any tips, keep to hear?

**Comment by vel21ripn** [comment]
We need to understand where the 64-bit division appears.
Previously, this problem appeared when converting 64-bit time to 32-bit.

I have just merged with the "dev" branch and am fixing some memory leaks during initialization.
In addition, I will need to fix issue #109

I hope to post an update by the end of the week.

**Comment by ewildgoose** [comment]
Was an easy find actually. It's an old issue re-occuring (looks similar to commit: 3bf4551e22 ). Fix is as follows:

```
--- src/lib/ndpi_main.c.orig	2021-04-12 12:20:13.441428167 +0000
+++ src/lib/ndpi_main.c	2021-04-12 12:21:47.407439031 +0000
@@ -5152,7 +5152,7 @@
 	if(ndpi_str->msteams_cache)
 	  ndpi_lru_add_to_cache(ndpi_str->msteams_cache,
 				flow->packet.iph->saddr,
-				(flow->packet.current_time_ms / 1000) & 0xFFFF /* 16 bit */);
+				flow->packet.current_time & 0xFFFF /* 16 bit */);
       }
       break;

@@ -5165,7 +5165,7 @@

 	if(ndpi_lru_find_cache(ndpi_str->msteams_cache, flow->packet.iph->saddr,
 			       &when, 0 /* Don't remove it as it can be used for other connections */)) {
-	  u_int16_t tdiff = ((flow->packet.current_time_ms /1000) & 0xFFFF) - when;
+	  u_int16_t tdiff = (flow->packet.current_time & 0xFFFF) - when;

 	  if(tdiff < 60 /* sec */) {
 	    // printf("====>> NDPI_PROTOCOL_SKYPE(_CALL) -> NDPI_PROTOCOL_MSTEAMS [%u]\n", tdiff);
@@ -5174,7 +5174,7 @@
 	    /* Refresh cache */
 	    ndpi_lru_add_to_cache(ndpi_str->msteams_cache,
 				  flow->packet.iph->saddr,
-				  (flow->packet.current_time_ms / 1000) & 0xFFFF /* 16 bit */);
+				  flow->packet.current_time & 0xFFFF /* 16 bit */);
 	  }
 	}
       }

```

Just a note for myself when I find myself here on google. I compiled this on 32bit x86, then did:
  objdump -Sr ndpi_main.o
And grep/scan for "udivdi3" 

I guess on arm I would do the same, but search for the other symbol.

**Comment by vel21ripn** [comment]
Thanks!
Fixed in 8364d516c210787cfb2baa9dcc988ca23d96a272
